api/tacticalrmm/core/utils.py

```diff
@@ -209,3 +214,201 @@ def get_meshagent_url(
 
 def make_alpha_numeric(s: str):
     return "".join(filter(str.isalnum, s))
+
+
+def find_and_replace_db_values_str(*, text: str, instance):
+    import re
+
+    from tacticalrmm.utils import RE_DB_VALUE, get_db_value
+
+    for string, model, prop in re.findall(RE_DB_VALUE, text):
+        value = get_db_value(string=f"{model}.{prop}", instance=instance)
+        return text.replace(string, str(value))
+    else:
+        return text
+
+
+def find_and_replace_db_values_dict(
+    *, dict_value: Dict[str, Any], instance
+) -> Dict[str, Any]:
+    new_dict = {}
+
+    for key, value in dict_value.items():
+        new_key = find_and_replace_db_values_str(text=key, instance=instance)
+
+        # Check if the value is a dictionary and recursively call the function if it is
+        if isinstance(value, dict):
+            new_value = find_and_replace_db_values_dict(
+                dict_value=value, instance=instance
+            )
+        else:
+            new_value = find_and_replace_db_values_str(text=value, instance=instance)
+
+        new_dict[new_key] = new_value
+
+    return new_dict
+
+
+def _run_url_rest_action(*, url: str, method, body: str, headers: str, instance=None):
+    # replace url
+    new_url = find_and_replace_db_values_str(text=url, instance=instance)
+    new_body = find_and_replace_db_values_dict(
+        dict_value=json.loads(body), instance=instance
+    )
+    new_headers = find_and_replace_db_values_dict(
+        dict_value=json.loads(headers), instance=instance
+    )
+
+    if method in ["get", "delete"]:
+        return getattr(requests, method)(new_url, headers=new_headers)
+    else:
+        return getattr(requests, method)(new_url, data=new_body, headers=new_headers)
+
+
+def run_url_rest_action(*, action_id: int, instance=None) -> Tuple[str, int]:
+    import core.models
+
+    action = core.models.URLAction.objects.get(pk=action_id)
+    method = action.rest_method
+    url = action.pattern
+    body = action.rest_body
+    headers = action.rest_headers
+
+    response = _run_url_rest_action(
+        url=url, method=method, body=body, headers=headers, instance=instance
+    )
+
+    return (response.text, response.status_code)
+
+
+def run_test_url_rest_action(
+    *,
+    url: str,
+    method,
+    body: str,
+    headers: str,
+    instance_type: Optional[str],
+    instance_id: Optional[int],
+) -> Tuple[str, int]:
+    lookup_instance = None
+    if instance_type and instance_id:
+        if instance_type == "client":
+            Client = apps.get_model("clients.Client")
```

Comment:
why are you using `app.get_model` here instead of importing Client?

Reply:
circular import issue

Reply:
reworked it to load it dynamically instead of a bunch of ifs